mod download issue

when trying to download the common mod tools mod i (and another user who reported the bug) receive the following error
```
java.util.concurrent.CompletionException: java.io.IOException: Server returned HTTP response code: 400 for URL: http://content.faforever.com/mods/Common Mod Tools.v0001.zip
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniRun.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)
	at com.faforever.client.task.CompletableTask.lambda$new$1(CompletableTask.java:17)
	at javafx.base/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(Unknown Source)
	at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(Unknown Source)
	at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(Unknown Source)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
	at javafx.base/com.sun.javafx.event.EventUtil.fireEventImpl(Unknown Source)
	at javafx.base/com.sun.javafx.event.EventUtil.fireEvent(Unknown Source)
	at javafx.base/javafx.event.Event.fireEvent(Unknown Source)
	at javafx.graphics/javafx.concurrent.EventHelper.fireEvent(Unknown Source)
	at javafx.graphics/javafx.concurrent.Task.fireEvent(Unknown Source)
	at javafx.graphics/javafx.concurrent.Task.setState(Unknown Source)
	at javafx.graphics/javafx.concurrent.Task$TaskCallable.lambda$call$2(Unknown Source)
	at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$9(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(Unknown Source)
	at javafx.graphics/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(Unknown Source)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.io.IOException: Server returned HTTP response code: 400 for URL: http://content.faforever.com/mods/Common Mod Tools.v0001.zip
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
	at java.base/java.lang.reflect.Constructor.newInstance(Unknown Source)
	at java.base/sun.net.www.protocol.http.HttpURLConnection$10.run(Unknown Source)
	at java.base/sun.net.www.protocol.http.HttpURLConnection$10.run(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getChainedException(Unknown Source)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source)
	at com.faforever.client.mod.InstallModTask.call(InstallModTask.java:66)
	at com.faforever.client.mod.InstallModTask.call(InstallModTask.java:32)
	at javafx.graphics/javafx.concurrent.Task$TaskCallable.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	... 1 more
Caused by: java.io.IOException: Server returned HTTP response code: 400 for URL: http://content.faforever.com/mods/Common Mod Tools.v0001.zip
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getHeaderField(Unknown Source)
	at java.base/java.net.URLConnection.getHeaderFieldLong(Unknown Source)
	at java.base/java.net.URLConnection.getContentLengthLong(Unknown Source)
	at java.base/java.net.URLConnection.getContentLength(Unknown Source)
	at com.faforever.client.mod.InstallModTask.call(InstallModTask.java:63)
	... 6 more
```
downloading the mod works when doing so via the http://content.faforever.com/mods/Common Mod Tools.v0001.zip or via the legacy client
i assume the client

--- a/src/main/java/com/faforever/client/config/JacksonConfig.java
+++ b/src/main/java/com/faforever/client/config/JacksonConfig.java
@@ -5,6 +5,8 @@ import com.fasterxml.jackson.databind.DeserializationContext;
 import com.fasterxml.jackson.databind.JsonDeserializer;
 import com.fasterxml.jackson.databind.Module;
 import com.fasterxml.jackson.databind.module.SimpleModule;
+import com.google.common.annotations.VisibleForTesting;
+
 import org.apache.maven.artifact.versioning.ComparableVersion;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
@@ -33,7 +35,8 @@ public class JacksonConfig {
     }
   }
 
-  private static class UrlDeserializer extends JsonDeserializer<URL> {
+  @VisibleForTesting
+  static class UrlDeserializer extends JsonDeserializer<URL> {
 
     @Override
     public URL deserialize(JsonParser p, DeserializationContext ctxt) throws IOException {

--- a/src/main/java/com/faforever/client/config/JacksonConfig.java
+++ b/src/main/java/com/faforever/client/config/JacksonConfig.java
@@ -10,6 +10,9 @@ import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 
 import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
 
 @Configuration
 public class JacksonConfig {
@@ -18,6 +21,7 @@ public class JacksonConfig {
   public Module customDeserializers() {
     SimpleModule fafModule = new SimpleModule();
     fafModule.addDeserializer(ComparableVersion.class, new ComparableVersionDeserializer());
+    fafModule.addDeserializer(URL.class, new UrlDeserializer());
     return fafModule;
   }
 
@@ -26,6 +30,21 @@ public class JacksonConfig {
     @Override
     public ComparableVersion deserialize(JsonParser p, DeserializationContext ctxt) throws IOException {
       return new ComparableVersion(p.getValueAsString());
+    }
+  }
+
+  private static class UrlDeserializer extends JsonDeserializer<URL> {
+
+    @Override
+    public URL deserialize(JsonParser p, DeserializationContext ctxt) throws IOException {
+      // the default URL parsing is bad, see https://stackoverflow.com/questions/10786042/java-url-encoding-of-query-string-parameters
+      URL url = new URL(p.getValueAsString());
+      try {
+        URI uri = new URI(url.getProtocol(), url.getUserInfo(), url.getHost(), url.getPort(), url.getPath(), url.getQuery(), url.getRef());
+        return uri.toURL();
+      } catch (URISyntaxException e) {
+        throw new IOException("Could not parse URL!", e);
+      }
     }
   }
 }

--- a/src/test/java/com/faforever/client/config/JacksonConfigTest.java
+++ b/src/test/java/com/faforever/client/config/JacksonConfigTest.java
@@ -1,0 +1,33 @@
+package com.faforever.client.config;
+
+import static org.junit.Assert.assertEquals;
+
+import java.io.IOException;
+import java.net.URL;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import com.faforever.client.config.JacksonConfig.UrlDeserializer;
+import com.fasterxml.jackson.core.JsonParser;
+
+@RunWith(MockitoJUnitRunner.class)
+public class JacksonConfigTest {
+
+  @Mock
+  JsonParser parser;
+
+  @Test
+  public void testUrlDeserializer() throws IOException {
+    String urlOriginal = "http://www.foo.bar/path with space/test?foo=arg with space&bar=baz";
+    String urlEncoded = "http://www.foo.bar/path%20with%20space/test?foo=arg%20with%20space&bar=baz";
+    
+    UrlDeserializer deserializer = new UrlDeserializer();
+    Mockito.when(parser.getValueAsString()).thenReturn(urlOriginal);
+    URL result = deserializer.deserialize(parser, null);
+    assertEquals(urlEncoded, result.toString());
+  }
+}

